### DISCUSSION TITLE
How to send url signal using custom activity

### DISCUSSION THREAD

**Discussion by yopichy** [elsa-workflows/elsa-core#2837] [Q&A]
I have a case of creating an approval business workflow with several levels of approval.
and I want to send a signal url using an existing mail server (API) instead of using elsa.

what I want to ask is how to get the signal url and use the http client with the token from custom activity and is it possible to use dynamic parameters for join mode every time it creates an instance?
any suggestion or example @sfmskywalker ?

![image](https://user-images.githubusercontent.com/2038089/158078854-33102b83-eafd-4eae-ad3a-7c860f69fdcf.png)

![image](https://user-images.githubusercontent.com/2038089/158078686-ebf513c0-b1a7-406f-ac55-f40f00c37a73.png)

**Comment by sfmskywalker** [comment]
What you should do is execute your API call as part of the workflow. Right now you're trying to do this from the `Build` method directly, but that only gets executed to build the activity. What you want though, is have your API call execute when the activity executes. For example:

```csharp
public override void Build(ICompositeActivityBuilder builder)
{
   builder.Then(async ctx => 
   {
      var reject = ctx.GenerateSignalUrl($"{TransactionId}_Reject");
      var client = new HttpClient();
      var response = await client.PostAsync("http://localhost:5151/api/notifyUser", reject);
      var outcome = response.IsSuccessStatusCode ? "Success" : "Failed";
      return new OutcomeResult(outcome);
   }, activity => {
      activity.When("Success")
      .Then<Fork>(...) // rest of your composite continues here.
   })
}
```

One small caveat: the inline activity API as demonstrated above does not currently seem to work due to a missing overload that allows you to return an outcome. I will fix that, but in the meantime you might just create a custom activity, like this:

```csharp
public class NotifyUser : Activity
{
   [ActivityInput] public string TrandactionId { get; set; }

   protected override async ValueTask<IActivityExecutionResult> OnExecuteAsync(ActiityExecutionContext ctx)
   {
      var reject = ctx.GenerateSignalUrl($"{TransactionId}_Reject");
      var client = new HttpClient();
      var response = await client.PostAsync("http://localhost:5151/api/notifyUser", reject);
      var outcome = response.IsSuccessStatusCode ? "Success" : "Failed";
      return Outcome(outcome);
   }
}
```

Then use it in your composite activity:

```csharp
public override void Build(ICompositeActivityBuilder builder)
{
   builder.Then<NotifyUser>(activity => activity.Set(x => x.TransactionId, TransactionId), activity => {
      activity.When("Success")
      .Then<Fork>(...) // rest of your composite continues here.
   })
}
```

**Reply by yopichy** [reply]
thanks for your explanation :)
i'll try it using your advise. thank you @sfmskywalker

**Reply by yopichy** [reply]
@sfmskywalker one more, how to get input payload from outcome to set param in http response?
![image](https://user-images.githubusercontent.com/2038089/159339911-115db63c-6b8c-4da9-a249-3aa04d42f4c7.png)

i want to pass input outcome to WriteHttpResponse, and after that waiting trigger from signalurl
![image](https://user-images.githubusercontent.com/2038089/159340088-7c64dce7-07e6-467e-97b1-59a2255a42fe.png)

i'm trying change code like this
![image](https://user-images.githubusercontent.com/2038089/159344952-36bd85b2-aaa3-49d4-8c97-9332a4ede872.png)
stop after exec generatesignalurl
![image](https://user-images.githubusercontent.com/2038089/159344804-eb43a1fe-9271-43d8-8d7d-d49622f69db8.png)

**Reply by sfmskywalker** [reply]
The way you access `ctx.Input` seem OK to me, it should be the output from `GenerateSignalUrl`.
One other thing you could try is to connect `GenerateSignalUrl` to `WriteHttpResponse` not from within the lambda as the second argument to the `Then<GenerateSignalUrl>` call, but "fluently", like this:

```csharp
.Then<GenerateSignalUrl>(act => act.Set(x => x.TransactionId, "Document")
.WriteHttpResponse(act => act.WithStatusCode(HttpStatusCode.OK).WithContent(ctx => ctx.Input))
.Then<ApprovalFork>(...)
..etc
```

Also keep an eye out on the console window in case of exceptions happening.
